### DISCUSSION TITLE
Learning about crypt14

### DISCUSSION THREAD

**Discussion by rahlhar** [ElDavoo/wa-crypt-tools#104] [General]
As an exercise, I'm implementing crypt14 decryption in `C`, using your excellent work as my guide.
I hit my first hurdle when I realised that the `key` extracted from a personal device turned out to be `158 bytes`, not the `131` that your code expects.

I took a chance looking at the `hexdump`, wondering if there is simply some extra `27 bytes`.
This does in fact appear to be the case - just prepended to the rest of the key.

I have at least partially confirmed this, by pulling out the bytes demarcated in [key14.py](https://github.com/ElDavoo/wa-crypt-tools/blob/main/src/wa_crypt_tools/lib/key/key14.py) as:

- googleIdSalt (unused?) (16 bytes)
- hashedGoogleID (The SHA-256 hash of googleIdSalt) (32 bytes)

and manually checking that the `sha256` hash of the id salt matches the subsequent hashed id bytes.

Do you have any idea what these initial `27 bytes` might be? Have you come across them before?
It looks like they can simply be ignored, but I just got to wondering.

**Comment by rahlhar** [comment]
[Here](https://github.com/aleksmarcelo/wadec) it seems to show both `crypt12` and `crypt14` as being `158 bytes`, with some unlabelled `30 bytes` at the beginning.
Have I perhaps overlooked something in your implementation?

**Comment by ElDavoo** [comment]
Hi, did you deserialize the Java object?

**Comment by rahlhar** [comment]
I guess the short answer is no.

Apologies, I'm not sure I follow.
Thus far I have only really begun digging through your code, trying to get an understanding of the stages involved.

I'll be honest and say that thus far I was avoiding installing `wa-crypt-tools` and simply using the source as reference.
I suppose I should though.

**Comment by rahlhar** [comment]
Ok, I see you reference these java bytes [here](https://github.com/ElDavoo/wa-crypt-tools/issues/16#issuecomment-1078836927).

I have obviously thus far missed that part within the source.

**Comment by ElDavoo** [comment]
This is why the javaobj-py3 library is used.

**Comment by rahlhar** [comment]
Shall continue to dig through - I somehow managed to overlook the use of `javaobj` in `keyfactory.py`.

**Comment by ElDavoo** [comment]
Yep, I guess documentation needs to be done

**Comment by rahlhar** [comment]
That's brilliant :) Happy to hear.

I'll be taking a non-object-oriented approach myself - not that one couldn't do the same with CPP.

I'm tempted to read that as `C Pre-Processor` because _that_ would be epic(ly hilarious).

**Comment by ElDavoo** [comment]
![image](https://github.com/ElDavoo/wa-crypt-tools/assets/7345120/42bf0521-d617-4e0e-8168-ae09336e9528)
